sycl/include/CL/sycl/accessor.hpp

```diff
@@ -723,83 +726,239 @@ class accessor<DataT, Dimensions, AccessMode, access::target::local,
 // Image accessor
 template <typename DataT, int Dimensions, access::mode AccessMode,
           access::target AccessTarget, access::placeholder IsPlaceholder>
-class image_accessor {
-  static_assert(AccessTarget == access::target::image ||
-                    AccessTarget == access::target::host_image ||
-                    AccessTarget == access::target::image_array,
+class image_accessor
+#ifndef __SYCL_DEVICE_ONLY__
+    : public detail::AccessorBaseHost {
+  size_t MImageSize;
+  size_t MImageCount;
+#else
+{
+  /*
+  // TODO: Define the datatype here based on Dimensions, AccessMode and
+  AccessTarget.
+  __ocl_image[Dim]d[array/non_array]_[AM]_t MImage;
+  __init(__ocl_imagexx_t Image) { MImage = Image; }
+  */
+#endif
+  constexpr static bool IsHostImageAcc =
+      (AccessTarget == access::target::host_image);
+
+  constexpr static bool IsImageAcc = (AccessTarget == access::target::image);
+
+  constexpr static bool IsImageArrayAcc =
+      (AccessTarget == access::target::image_array);
+
+  constexpr static bool IsImageAccessAnyWrite =
+      (AccessMode == access::mode::write ||
+       AccessMode == access::mode::discard_write);
+
+  constexpr static bool IsImageAccessRead = (AccessMode == access::mode::read);
+
+  static_assert(IsImageAcc || IsHostImageAcc || IsImageArrayAcc,
                 "Expected image type");
-  // TODO: Check if placeholder is applicable here.
+
+  static_assert(IsPlaceholder == access::placeholder::false_t,
+                "Expected false as Placeholder value for image accessor.");
+
+  static_assert(
+      AccessMode == access::mode::read || AccessMode == access::mode::write ||
+          AccessMode == access::mode::discard_write,
+      "Access modes can be only read/write/discard_write for image accessor.");
+
+  static_assert(Dimensions > 0 && Dimensions <= 3,
+                "Dimensions can be 1/2/3 for image accessor.");
+
 public:
   using value_type = DataT;
   using reference = DataT &;
   using const_reference = const DataT &;
 
-  /* Available only when: accessTarget == access::target::host_image */
+  // image_accessor Constructors.
+
+  // Available only when: accessTarget == access::target::host_image
   // template <typename AllocatorT>
   // accessor(image<dimensions, AllocatorT> &imageRef);
-  /* Available only when: accessTarget == access::target::image */
+  template <typename AllocatorT, int Dims = Dimensions,
+            typename = detail::enable_if_t<(Dims > 0 && Dims <= 3) && IsHostImageAcc>>
+  image_accessor(image<Dims, AllocatorT> &ImageRef, int ImageElementSize)
+#ifdef __SYCL_DEVICE_ONLY__
+  = delete;
```

Comment:
I think general approach is to use SFINAE to make some methods unavailable.
Why we use different approach here?

Reply:
Different approach to diasble the constructor for SYCL_DEVICE_ONLY?

Reply:
Yes. 
I would be nice to keep the code consistent.
If new approach has benefits over old one, we should apply it to buffer accessors also, but I'm not sure I understand the motivation.

Reply:
I tested my code and found that it is incorrect to use the delete keyword at that location.
The use of delete is to emphasize that the constructor cannot be called in the device kernel code.

However, the whole code , including the non-kernel code is parsed by the device compiler with SYCL_DEVICE_ONLY macro set. The compiler sees that we are trying to call the constructor which has been marked as deleted, and gives compilation error; although the call is outside the kernel code.

I think this is why delete was not used in the buffer/local accessor constructors, and they have some definitions for the constructors with SYCL_DEVICE_ONLY set, even though the code is never executed.

I will upload the fix.

Thanks.